### PR TITLE
Changed Unit Of Work instead Repositories in Managers and refactored related items

### PR DESCRIPTION
**Pull request by zinkoviktor**
1. In all pojects:
1.1. Change ICollection to IEnumarable
1.2. Refactor all model, entities etc to have Collections in bottom
2. In DataLayer project:
2.1. In IUnitOfWork: Create method change to return IEnumarable, Update and Delete methods change to void
2.2. IDbContext rename to IUnitOfWorkDbContext
2.3. In BaseUnitOfWork change all methods as virtual
3. In BusinessLayer project:
3.1. Create generic BaseManager, inject IUnitOfWork implement all methods such as in IUnitOfWork
3.2. In all menagers change IRepository to IUnitOfWork inherit all managers from BaseManager.
4. In DataLayer.EF project:
4.1. Remove all mocks
4.2. IRepositoryDbContext rename to IDbContext
4.3. Remane EfDbContext to ProductsDbContext
4.4. In OnModel creating specify fields length and type.
4.5. Mock creating models in ProductsDbContext
5.6. In IRepository: Create method change to return IEnumarable, Update and Delete methods change to void
5. In WebApi project:
5.1. To implement next layout: 1 - name => 2 - price, 3 - available count, 4 - description, for get all products response
5.2. Set returning all products in browser after starting application

### CODE REVIEW

File: NetCoreRestApi/DataLayer.EF/Entities/ProductEntity.cs [hunk 1]
```diff
@@ -3,9 +3,9 @@
 namespace DataLayer.EF.Entities
 {
     public class ProductEntity : ItemEntity<int>
-    {      
-        public ICollection<CategoryEntity> Categories { get; set; }
+    {
         public decimal Price { get; set; }
         public int AvailableCount { get; set; }
+        public IEnumerable<CategoryEntity> Categories { get; set; }
```
Comment (mdemii) [inline NetCoreRestApi/DataLayer.EF/Entities/ProductEntity.cs]: Revert this change because this is entity so it should be ICollection

File: NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+using Microsoft.EntityFrameworkCore;
+using System.Collections.Generic;
+
+namespace DataLayer.EF
+{
+    public class ProductsDbContext : DbContext, IDbContext       
+    {
+        public ProductsDbContext(DbContextOptions<ProductsDbContext> options) : base(options)
+        {
+            AddRange(GetProductsMock());
```
Comment (mdemii) [inline NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs]: This code shouldn't be here as it is mock data. So create ProductsMockDbContext that inherits ProductDbContext and move seeding code into that class and configure DI to use this class instead of ProductDbContext

File: NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                ep.Property(p => p.Description).HasMaxLength(150);
+            });
+
+            modelBuilder.Entity<CategoryEntity>(ec =>
+            {
+                ec.HasKey(c => c.Id);
+                ec.Property(c => c.Id).IsRequired();
+                ec.Property(c => c.Name).HasMaxLength(50);
+                ec.Property(c => c.Description).HasMaxLength(150);
+            });
```
Comment (mdemii) [inline NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs:34]: Specify relationship between category and product tables here

File: NetCoreRestApi/DataLayer/Repositories/Interfaces/IUnitOfWorkDbContext.cs [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace DataLayer.Repositories.Intefaces
 {
-    public interface IDbContext
+    public interface IUnitOfWorkDbContext
```
Comment (mdemii) [inline NetCoreRestApi/DataLayer/Repositories/Interfaces/IUnitOfWorkDbContext.cs]: Rename to IUnitOfWorkContext

File: NetCoreRestApi/WebAPI/Startup.cs [hunk 5]
```diff
@@ -30,10 +33,15 @@ public void ConfigureServices(IServiceCollection services)
...
+            services.AddTransient<ICategoryUnitOfWork, CategoryUnitOfWork>();
+            services.AddTransient<IProductUnitOfWork, ProductUnitOfWork>();
+            services.AddTransient<IConverter<CategoryEntity, CategoryModel>, CategoryModelConverter>();
             services.AddTransient<ICategoryRepository, CategoryRepository>();
             services.AddTransient<IProductRepository, ProductRepository>();
-            services.AddTransient<IRepositoryDbContext, EfDbContext>();
-            services.AddDbContext<EfDbContext>(opt => opt.UseInMemoryDatabase(databaseName: Guid.NewGuid().ToString()), ServiceLifetime.Singleton, ServiceLifetime.Singleton);
+            services.AddTransient<IUnitOfWorkDbContext, ProductsDbContext>();
+            services.AddTransient<IDbContext, ProductsDbContext>();
+            services.AddDbContext<ProductsDbContext>(opt => opt.UseInMemoryDatabase(databaseName: Guid.NewGuid().ToString()),
```
Comment (mdemii) [inline NetCoreRestApi/WebAPI/Startup.cs:43]: Use here ProductMockDbContext type

**Review by mdemii — CHANGES_REQUESTED** [review]
Changes are required

File: NetCoreRestApi/DataLayer.EF/Entities/ProductEntity.cs (hunk 1 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/Entities/ProductEntity.cs]: fixed

File: NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs (hunk 2 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs]: fixed

File: NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs (hunk 3 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/ProductsDbContext.cs:34]: should discus

File: NetCoreRestApi/DataLayer/Repositories/Interfaces/IUnitOfWorkDbContext.cs (hunk 4 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer/Repositories/Interfaces/IUnitOfWorkDbContext.cs]: fixed

File: NetCoreRestApi/WebAPI/Startup.cs (hunk 5 above)
Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Startup.cs:43]: fixed
